core/cmd/shell_local.go

```diff
@@ -1069,41 +1029,138 @@ func (s *Shell) RemoveBlocks(c *cli.Context) error {
 	}
 
 	lggr := logger.Sugared(s.Logger.Named("RemoveBlocks"))
-	ldb := pg.NewLockedDB(cfg.AppID(), cfg.Database(), cfg.Database().Lock(), lggr)
 	ctx, cancel := context.WithCancel(context.Background())
 	go shutdown.HandleShutdown(func(sig string) {
 		cancel()
 		lggr.Info("received signal to stop - closing the database and releasing lock")
 
-		if cErr := ldb.Close(); cErr != nil {
-			lggr.Criticalf("Failed to close LockedDB: %v", cErr)
+		s.afterNode(lggr)
+	})
+
+	app, err := s.AppFactory.NewApplication(ctx, s.Config, s.Logger, s.Registerer, s.DS, s.KeyStore)
+	if err != nil {
+		return s.errorOut(errors.Wrap(err, "fatal error instantiating application"))
+	}
+
+	err = app.DeleteLogPollerDataAfter(ctx, chainID, start)
+	if err != nil {
+		return s.errorOut(err)
+	}
+
+	lggr.Infof("RemoveBlocks: successfully removed blocks")
+
+	return nil
+}
+
+// BeforeNode initializes database, keystore, logger, and beholder for node startup.
+// This is used as a Before hook in CLI commands that require these components.
+func (s *Shell) BeforeNode(c *cli.Context) error {
+	if err := s.beforeNode(c); err != nil {
+		return s.errorOut(err)
+	}
+	return nil
+}
+
+// afterNode is a thread-safe helper method to close the database and logger.
+// This is used in multiple places: shutdown handler, signal handler, and cleanup.
+// It uses sync.Once to ensure cleanup happens exactly once, even if called concurrently.
+func (s *Shell) afterNode(lggr logger.SugaredLogger) {
+	s.CleanupOnce.Do(func() {
+		if err := s.LDB.Close(); err != nil {
+			lggr.Criticalf("Failed to close LockedDB: %v", err)
 		}
+		lggr.Debug("Closed DB")
 
-		if cErr := s.CloseLogger(); cErr != nil {
-			log.Printf("Failed to close Logger: %v", cErr)
+		if err := s.CloseLogger(); err != nil {
+			log.Printf("Failed to close Logger: %v", err)
 		}
+		log.Printf("Closed logger")
```

Comment:
I think we should be minimizing these logs changes. Is there some reason these new logs lines are important?
```suggestion

		if err := s.CloseLogger(); err != nil {
			log.Printf("Failed to close Logger: %v", err)
}
```

Reply:
Done